File: app/backend/schemas/proposta_pc.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PropostaPcMaoObraOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    descricao_funcao: str
    codigo_origem: str | None
    quantidade: Decimal | None
    salario: Decimal | None
    previsao_reajuste: Decimal | None
    encargos_percent: Decimal | None
    periculosidade_insalubridade: Decimal | None
    refeicao: Decimal | None
    agua_potavel: Decimal | None
    vale_alimentacao: Decimal | None
    plano_saude: Decimal | None
    ferramentas_val: Decimal | None
    seguro_vida: Decimal | None
    abono_ferias: Decimal | None
    uniforme_val: Decimal | None
    epi_val: Decimal | None
    custo_unitario_h: Decimal | None
    custo_mensal: Decimal | None
    mobilizacao: Decimal | None
    valor_bcu_snapshot: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcEquipamentoPremissaOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    horas_mes: Decimal | None
    preco_gasolina_l: Decimal | None
    preco_diesel_l: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcEquipamentoOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    codigo: str | None
    codigo_origem: str | None
    equipamento: str
    combustivel_utilizado: str | None
    consumo_l_h: Decimal | None
    aluguel_r_h: Decimal | None
    combustivel_r_h: Decimal | None
    mao_obra_r_h: Decimal | None
    hora_produtiva: Decimal | None
    hora_improdutiva: Decimal | None
    mes: Decimal | None
    aluguel_mensal: Decimal | None
    valor_bcu_snapshot: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcEncargoOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    tipo_encargo: str
    grupo: str | None
    codigo_grupo: str | None
    discriminacao_encargo: str
    taxa_percent: Decimal | None
    valor_bcu_snapshot: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcEpiOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    codigo_origem: str | None
    epi: str
    unidade: str | None
    custo_unitario: Decimal | None
    quantidade: Decimal | None
    vida_util_meses: Decimal | None
    custo_epi_mes: Decimal | None
    valor_bcu_snapshot: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcFerramentaOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    codigo_origem: str | None
    item: str | None
    descricao: str
    unidade: str | None
    quantidade: Decimal | None
    preco: Decimal | None
    preco_total: Decimal | None
    valor_bcu_snapshot: Decimal | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class PropostaPcMobilizacaoOut(BaseModel):
    id: UUID
    proposta_id: UUID
    bcu_item_id: UUID | None
    descricao: str
    funcao: str | None
    tipo_mao_obra: str | None
    editado_manualmente: bool
    model_config = ConfigDict(from_attributes=True)


class DivergenciaOut(BaseModel):
    tabela: Literal["mao-obra", "equipamento", "epi", "ferramenta", "encargo"]
    item_id: UUID
    campo: str
    valor_snapshot: Decimal | None
    valor_atual_bcu: Decimal | None
    valor_proposta: Decimal | None


class RecursoExtraOut(BaseModel):
    id: UUID
    proposta_id: UUID
    tipo_recurso: str
    descricao: str
    unidade_medida: str | None
    custo_unitario: Decimal
    observacao: str | None
    alocacoes_count: int


class HistogramaCompletoResponse(BaseModel):
    proposta_id: UUID
    bcu_cabecalho_id: UUID | str | None
    mao_obra: list[PropostaPcMaoObraOut]
    equipamento_premissa: PropostaPcEquipamentoPremissaOut | None
    equipamentos: list[PropostaPcEquipamentoOut]
    encargos_horista: list[PropostaPcEncargoOut]
    encargos_mensalista: list[PropostaPcEncargoOut]
    epis: list[PropostaPcEpiOut]
    ferramentas: list[PropostaPcFerramentaOut]
    mobilizacao: list[PropostaPcMobilizacaoOut]
    recursos_extras: list[RecursoExtraOut]
    divergencias: list[DivergenciaOut]
    cpu_desatualizada: bool


class MontarHistogramaResponse(BaseModel):
    mao_obra: int
    equipamento_premissa: int
    equipamentos: int
    encargos: int
    epis: int
    ferramentas: int
    mobilizacao: int


class RecursoExtraCreate(BaseModel):
    tipo_recurso: str
    descricao: str
    unidade_medida: str | None = None
    custo_unitario: Decimal
    observacao: str | None = None


class RecursoExtraUpdate(BaseModel):
    descricao: str | None = None
    unidade_medida: str | None = None
    custo_unitario: Decimal | None = None
    observacao: str | None = None


class AlocarRecursoRequest(BaseModel):
    recurso_extra_id: UUID
    quantidade_consumo: Decimal = Decimal("1")


class AlocacaoOut(BaseModel):
    id: UUID
    recurso_extra_id: UUID
    composicao_id: UUID
    quantidade_consumo: Decimal
    model_config = ConfigDict(from_attributes=True)

File: app/frontend/src/shared/types/contracts/proposta_pc.ts
export interface PropostaPcMaoObraOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  descricao_funcao: string;
  codigo_origem: string | null;
  quantidade: number | null;
  salario: number | null;
  previsao_reajuste: number | null;
  encargos_percent: number | null;
  periculosidade_insalubridade: number | null;
  refeicao: number | null;
  agua_potavel: number | null;
  vale_alimentacao: number | null;
  plano_saude: number | null;
  ferramentas_val: number | null;
  seguro_vida: number | null;
  abono_ferias: number | null;
  uniforme_val: number | null;
  epi_val: number | null;
  custo_unitario_h: number | null;
  custo_mensal: number | null;
  mobilizacao: number | null;
  valor_bcu_snapshot: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcEquipamentoPremissaOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  horas_mes: number | null;
  preco_gasolina_l: number | null;
  preco_diesel_l: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcEquipamentoOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  codigo: string | null;
  codigo_origem: string | null;
  equipamento: string;
  combustivel_utilizado: string | null;
  consumo_l_h: number | null;
  aluguel_r_h: number | null;
  combustivel_r_h: number | null;
  mao_obra_r_h: number | null;
  hora_produtiva: number | null;
  hora_improdutiva: number | null;
  mes: number | null;
  aluguel_mensal: number | null;
  valor_bcu_snapshot: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcEncargoOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  tipo_encargo: string;
  grupo: string | null;
  codigo_grupo: string | null;
  discriminacao_encargo: string;
  taxa_percent: number | null;
  valor_bcu_snapshot: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcEpiOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  codigo_origem: string | null;
  epi: string;
  unidade: string | null;
  custo_unitario: number | null;
  quantidade: number | null;
  vida_util_meses: number | null;
  custo_epi_mes: number | null;
  valor_bcu_snapshot: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcFerramentaOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  codigo_origem: string | null;
  item: string | null;
  descricao: string;
  unidade: string | null;
  quantidade: number | null;
  preco: number | null;
  preco_total: number | null;
  valor_bcu_snapshot: number | null;
  editado_manualmente: boolean;
}

export interface PropostaPcMobilizacaoOut {
  id: string;
  proposta_id: string;
  bcu_item_id: string | null;
  descricao: string;
  funcao: string | null;
  tipo_mao_obra: string | null;
  editado_manualmente: boolean;
}

export interface DivergenciaOut {
  tabela: 'mao-obra' | 'equipamento' | 'epi' | 'ferramenta' | 'encargo';
  item_id: string;
  campo: string;
  valor_snapshot: number | null;
  valor_atual_bcu: number | null;
  valor_proposta: number | null;
}

export interface RecursoExtraOut {
  id: string;
  proposta_id: string;
  tipo_recurso: string;
  descricao: string;
  unidade_medida: string | null;
  custo_unitario: number;
  observacao: string | null;
  alocacoes_count: number;
}

export interface HistogramaCompletoResponse {
  proposta_id: string;
  bcu_cabecalho_id: string | null;
  mao_obra: PropostaPcMaoObraOut[];
  equipamento_premissa: PropostaPcEquipamentoPremissaOut | null;
  equipamentos: PropostaPcEquipamentoOut[];
  encargos_horista: PropostaPcEncargoOut[];
  encargos_mensalista: PropostaPcEncargoOut[];
  epis: PropostaPcEpiOut[];
  ferramentas: PropostaPcFerramentaOut[];
  mobilizacao: PropostaPcMobilizacaoOut[];
  recursos_extras: RecursoExtraOut[];
  divergencias: DivergenciaOut[];
  cpu_desatualizada: boolean;
}

export interface MontarHistogramaResponse {
  mao_obra: number;
  equipamento_premissa: number;
  equipamentos: number;
  encargos: number;
  epis: number;
  ferramentas: number;
  mobilizacao: number;
}

export interface RecursoExtraCreate {
  tipo_recurso: string;
  descricao: string;
  unidade_medida?: string | null;
  custo_unitario: number;
  observacao?: string | null;
}

export interface RecursoExtraUpdate {
  descricao?: string | null;
  unidade_medida?: string | null;
  custo_unitario?: number | null;
  observacao?: string | null;
}

export interface AlocarRecursoRequest {
  recurso_extra_id: string;
  quantidade_consumo: number;
}

export interface AlocacaoOut {
  id: string;
  recurso_extra_id: string;
  composicao_id: string;
  quantidade_consumo: number;
}
